### PR TITLE
8322383: G1: Only preserve marks on objects that are actually moved

### PR DESCRIPTION
**Pull request by rkennke**
The G1 full-GC preserves marks during marking, for all live objects in compaction region. However, not all live objects do actually move. In particular, the start of a compaction chain may have a sediment of all-live objects which would not move, and thus don't need to have their marks preserved.
The problem can easily be solved by preserving marks during forwarding. That also seems a more natural place to do that.

Testing:
 - [x] hotspot_gc
 - [x] tier1
 - [ ] tier2

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8322383](https://bugs.openjdk.org/browse/JDK-8322383): G1: Only preserve marks on objects that are actually moved (**Enhancement** - P4)

### Reviewers
 * [Albert Mingkun Yang](https://openjdk.org/census#ayang) (@albertnetymk - **Reviewer**)
 * [Thomas Schatzl](https://openjdk.org/census#tschatzl) (@tschatzl - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/17159/head:pull/17159` \
`$ git checkout pull/17159`

Update a local copy of the PR: \
`$ git checkout pull/17159` \
`$ git pull https://git.openjdk.org/jdk.git pull/17159/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 17159`

View PR using the GUI difftool: \
`$ git pr show -t 17159`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/17159.diff">https://git.openjdk.org/jdk/pull/17159.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/17159#issuecomment-1863086821)

### CODE REVIEW

File: src/hotspot/share/gc/g1/g1FullGCCompactionPoint.cpp [hunk 1]
```diff
@@ -102,6 +103,7 @@ void G1FullGCCompactionPoint::forward(oop object, size_t size) {
 
   // Store a forwarding pointer if the object should be moved.
   if (cast_from_oop<HeapWord*>(object) != _compaction_top) {
+    preserved_stack()->push_if_necessary(object, object->mark());
```
Comment (albertnetymk) [inline src/hotspot/share/gc/g1/g1FullGCCompactionPoint.cpp:106]: Can this be made conditionally on whether the markword is NOT marked/forwarded? (IOW, move the added predicate in `markWord` here.) The rationale is to minimize changes to the shared code for G1 specific usages.

Comment (albertnetymk) [reply src/hotspot/share/gc/g1/g1FullGCCompactionPoint.cpp:107]: My previous msg was probably unclear... I meant sth like:

```c
if (!object->is_forwarded) {
  preserved_stack()->push_if_necessary...
}
```

(The newly added assert here is incorrect -- fails in GHA.)

**Review by albertnetymk — APPROVED** [review]

File: src/hotspot/share/gc/g1/g1FullGCCompactionPoint.cpp (hunk 1 above)
Comment (rkennke) [reply src/hotspot/share/gc/g1/g1FullGCCompactionPoint.cpp]: Oh I see! This makes more sense, yes. I made that change and cleaned up non G1 code.

BTW: I am working on new/alternative full-GCs for Serial, G1 and Shenandoah that don't store forwarding pointers in the mark-word at all, and thus avoid all that preserved-headers stuff (and together with the OM-world work that some Oracle engineers are working on, get rid of header displacement altogether). The SerialGC version is almost ready in Lilliput: https://github.com/openjdk/lilliput/pull/122

**Review by tschatzl — APPROVED** [review]

**Pull request closed without merging**